community: CrateDB: Vector Store

amotl opened the pull request:
### About
  - **Description:** Vector Store adapter for CrateDB.
  - **Coming from:** https://github.com/crate-workbench/langchain/pull/1
  - **Documentation:** GH-27713
  - **Addressed to:** @eyurtsev

### Status
- We are considering the patch ready for review and merging, with a few spots to be handled on a later iteration.
- Please let us know if you want to see any other details to be addressed before the initial merge.
- A few backlog items have been collected here: https://github.com/crate-workbench/langchain/issues/30.

### Sandbox
A little walkthrough how to exercise the software tests on your workstation.
```shell
docker run --rm -it --name=cratedb \
  --publish=4200:4200 --publish=5432:5432 --env=CRATE_HEAP_SIZE=2g \
  crate:latest -Cdiscovery.type=single-node
```
```shell
git clone https://github.com/crate-workbench/langchain.git --branch=cratedb-up/1/vector-store
cd langchain
uv venv
source .venv/bin/activate
cd libs/community
uv pip install --upgrade --prerelease=allow --editable=. poetry sqlalchemy-cratedb
poetry install --no-interaction --no-ansi --with dev,test,test_integration
```
```shell
pytest -vvv tests/integration_tests/vectorstores/test_cratedb.py
```

### Trivia
The CrateDB implementation is heavily based on PGVector's, with a few adjustments. Previous generalizations and improvements to PGVector have been submitted the other day already.

- GH-16243
- GH-16244

amotl committed:
CrateDB: Vector Store

amotl commented on libs/community/langchain_community/vectorstores/__init__.py:
```diff
@@ -334,6 +337,7 @@
     "Clickhouse",
     "ClickhouseSettings",
     "CouchbaseVectorStore",
+    "CrateDBVectorSearch",
```
Last call about »naming things« matters: `CrateDBVectorSearch` vs. `CrateDBVectorStore`?

Current cardinalities, amongst implementations from other vendors:
- VectorSearch: 9
- VectorStore: 9

I am tending to rename the class to `CrateDBVectorStore`, what do others think about it?

amotl replied:
The classes have been renamed to `CrateDBVectorStore` resp. `CrateDBVectorStoreMultiCollection`.

amotl commented on libs/community/langchain_community/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,453 @@
+    return [doc for doc, _ in docs_and_scores]
+
+
+class CrateDBVectorSearch(PGVector):
+    """`CrateDB` vector store.
```
It is PGVector, mostly, with a few adjustments. Previous generalizations to make this fly have been submitted the other day already.

- GH-16243
- GH-16244

amotl commented on libs/community/langchain_community/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,453 @@
+class CrateDBVectorSearch(PGVector):
+    """`CrateDB` vector store.
+
+    To use it, you should have the ``crate[sqlalchemy]`` python package installed.
```
Well, time flies. Please adjust package names.
```suggestion
    To use it, you should have the ``sqlalchemy-cratedb`` python package installed.
```

amotl replied:
Fixed per ffda5c8853f.

amotl commented on libs/community/langchain_community/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,453 @@
+                "CrateDBVectorSearch constructor."
+            )
+
+    @staticmethod
+    def _euclidean_relevance_score_fn(score: float) -> float:
+        """Return a similarity score on a scale [0, 1]."""
+        # The 'correct' relevance function
+        # may differ depending on a few things, including:
+        # - the distance / similarity metric used by the VectorStore
+        # - the scale of your embeddings (OpenAI's are unit normed. Many
+        #  others are not!)
+        # - embedding dimensionality
+        # - etc.
+        # This function converts the euclidean norm of normalized embeddings
+        # (0 is most similar, sqrt(2) most dissimilar)
+        # to a similarity function (0 to 1)
+
+        # Original:
+        # return 1.0 - distance / math.sqrt(2)
+        return score / math.sqrt(2)
```
Please review this.

/cc @ckurze, @kneth, @surister

amotl replied:
Are those matters relevant and applicable here?

- https://github.com/crate/crate/issues/15835
- https://github.com/crate/crate/pull/15832

kneth replied:
Isn't `_score` based on BM25? If so,  is `1 - score` or `1 / score` closer to the API documentation?

amotl replied:
> Isn't `_score` based on BM25?

I think it is some sort of hybrid. @surister or @matriv may be able to elaborate further?

> If so, is `1 - score` or `1 / score` closer to the API documentation?

I also would like to defer this question to subject matter experts. ;]

amotl replied:
If someone wants to look into this, this guideline may help to exercise relevant code paths that leverage the `_euclidean_relevance_score_fn()` method.

It is those two test cases that will:
```python
tests/integration_tests/vectorstores/test_cratedb.py::test_cratedb_relevance_score
tests/integration_tests/vectorstores/test_cratedb.py::test_cratedb_retriever_search_threshold
```

You can invoke them on the spot, exclusively, by using this pytest command:
```shell
pytest -vvv tests/integration_tests/vectorstores/test_cratedb.py -k "test_cratedb_relevance_score or test_cratedb_retriever_search_threshold"
```

surister replied:
Our full-text search function `match` gives a `_score` , which is the result of lucene's implementation of bm25. We also have `knn_match` and `vector_similarity` giving us a `_score`, even though then have they same name, they are different things.

BaurzhanSakhariev replied:
Is the goal here to return score normalized to a unit interval (0,1] where score itself based on euclidean distance?

As @surister mentioned, _score in fulltext is a different thing.

In CrateDB, both `knn_match` and `vector_similarity` use Lucene's `VectorSimilarityFunction.EUCLIDEAN` 
which is computed as `1 / (1 + squareDistance(v1, v2))`

We introduced  `vector_similarity` to get "pure" similarity value since `knn_match` can sometimes return _score outside of (0,1] interval, see https://github.com/crate-workbench/langchain/issues/19. It happens because Lucene's engine may take into account some other things when comparing docs which affect the final score.


1. I think parameter here must be called `distance`, at least in another repo `_euclidean_relevance_score_fn` 
   looks different https://github.com/crate-workbench/langchain/blob/cratedb/libs/core/langchain_core/vectorstores/base.py#L409-L421. 
   
 If physical meaning here is indeed score and not the distance, then it's either normalized and we don't need to do anything or it's affected by something else like in  https://github.com/crate-workbench/langchain/issues/19 and we can't divide by sqrt(2) as we don't know exactly what happened to the original score.

2. In case we use https://github.com/crate-workbench/langchain/blob/cratedb/libs/core/langchain_core/vectorstores/base.py#L409-L421. - it looks unaligned with CrateDB's `vector_similarity` both in docs and implementation.

amotl replied:
Hi. Thanks for your suggestions. I've responded to them [here](https://github.com/langchain-ai/langchain/pull/27710#discussion_r1832460771).

amotl replied:
Based on your suggestions, the patch has been adjusted with 7ff13192b1f. Thanks!

amotl commented on libs/community/extended_testing_deps.txt:
```diff
@@ -14,6 +14,7 @@ chardet>=5.1.0,<6
 cloudpathlib>=0.18,<0.19
 cloudpickle>=2.0.0
 cohere>=4,<6
+crate==1.0.0.dev1
```
### About
Last summer, we removed the SQLAlchemy dialect from the fundamental HTTP driver package. The next release of the [crate](https://github.com/crate/crate-python) package, version 1.0.0, will conclude the transition. Afterwards, it will be fine to just pull in the new [sqlalchemy-cratedb](https://github.com/crate/sqlalchemy-cratedb) package.

### References
- https://github.com/crate/crate-python/pull/616
- https://github.com/crate/crate-python/issues/620

amotl replied:
We have been doing some yak shaving towards a 1.0.0 release.

- https://github.com/crate-workbench/langchain/issues/29

amotl replied:
Version 1.0.0 of crate-python has been released, e6ccd807bb updates this branch accordingly.

amotl commented on libs/community/langchain_community/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,453 @@
+
+            _type = self.EmbeddingStore
+
+            results: List[Any] = (
+                session.query(  # type: ignore[attr-defined]
+                    self.EmbeddingStore,
+                    # TODO: Original pgvector code uses `self.distance_strategy`.
+                    #       CrateDB currently only supports EUCLIDEAN.
+                    #       self.distance_strategy(embedding).label("distance")
+                    sqlalchemy.literal_column(
+                        f"{self.EmbeddingStore.__tablename__}._score"
+                    ).label("_score"),
+                )
+                .filter(filter_by)
+                # CrateDB applies `KNN_MATCH` within the `WHERE` clause.
+                .filter(
+                    sqlalchemy.func.knn_match(
+                        self.EmbeddingStore.embedding, embedding, k
+                    )
+                )
+                .order_by(sqlalchemy.desc("_score"))
+                .join(
+                    self.CollectionStore,
+                    self.EmbeddingStore.collection_id == self.CollectionStore.uuid,
+                )
+                .limit(k)
+            )
```
I guess it is just NOT ok to use `_score` here, and this has just been applied in the interim?
Isn't this the spot where that feature request was coming from, to be able to use the actual vector distance?

- https://github.com/crate/crate/issues/15835

/cc @ckurze

amotl replied:
I dearly hope so!

amotl replied:
In the best case, just using a corresponding SQLAlchemy incantation (correctly) will work without much ado.

```python
import sqlalchemy as sa

sa.func.vector_similarity(...)
```

amotl replied:
Swapping in.
```python
sqlalchemy.func.vector_similarity(
    self.EmbeddingStore.embedding, embedding
).label("_score"),
```

Not there yet.
- https://github.com/crate/crate/issues/16912

amotl replied:
Fixed per 476d718b998e, with improvements pending for a later iteration.

surister replied:
I'm sorry but I don't understand the problem, what is it that we want? vector distance? if so, we do no directly have it, our `vector_similarity` gives us similarity based on euclidian distance, similarity is calculated like `sim = 1 / distance`

amotl replied:
As per [helper function above](https://github.com/langchain-ai/langchain/pull/27710#discussion_r1821498985), I think `similarity / math.sqrt(2)` will be used for normalization.

BaurzhanSakhariev replied:
Relates to https://github.com/langchain-ai/langchain/pull/27710/#discussion_r1832222614, +1 to `vector_similarity` and there is no need to normalize it. 

If https://github.com/crate/crate/issues/16912 was a blocker, 5.9.3 will have it fixed.

amotl replied:
It's not a blocker, we have been able to work around it. Thanks.

amotl committed:
CrateDB: Vector Store -- rename to CrateDBVectorStore

amotl committed:
CrateDB: Vector Store -- improve inline documentation

amotl committed:
CrateDB: Vector Store -- make it work using CrateDB's vector_similarity

Before, the adapter used CrateDB's built-in `_score` field for ranking.
Now, it uses the dedicated `vector_similarity()` function to compute the
similarity between two vectors.

surister reviewed (approved):
LGTM

amotl committed:
CrateDB: Vector Store -- update dependencies

amotl commented on libs/community/langchain_community/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,475 @@
+                "CrateDBVectorStore constructor."
+            )
+
+    @staticmethod
+    def _euclidean_relevance_score_fn(similarity: float) -> float:
+        """Return a similarity score on a scale [0, 1]."""
+        # The 'correct' relevance function
+        # may differ depending on a few things, including:
+        # - the distance / similarity metric used by the VectorStore
+        # - the scale of your embeddings (OpenAI's are unit normed. Many
+        #  others are not!)
+        # - embedding dimensionality
+        # - etc.
+        # This function converts the euclidean norm of normalized embeddings
+        # (0 is most similar, sqrt(2) most dissimilar)
+        # to a similarity function (0 to 1)
+
+        # Original:
+        # return 1.0 - distance / math.sqrt(2)
+        return similarity / math.sqrt(2)
```
Hi @BaurzhanSakhariev. Thanks for your reply above. I am all in for better naming things. Compared to the original variant still visible on the PR, the argument name already had been renamed to `similarity`, as I assumed this is the kind/unit of value that will be emitted from CrateDB's `vector_similarity()` function.

Please advise if it should be renamed once more, like ... as you suggested.

BaurzhanSakhariev replied:
Ah, so similarity argument here is a value returned from CrateDB?

Then we don't need anything on top of it, ie we don't need this function and instead should use value from CrateDB as is.

Similarity is already in the (0,1] interval and dividing by math.sqrt(2) won't normalize it but return wrong result, for example 1 will become  0.714.

amotl replied:
Thank you very much. The patch has been adjusted with 7ff13192b1f.

BaurzhanSakhariev replied:
Thanks!

Just checking the commit, looks like we had exactly one instance of
>for example 1 will become 0.714.

```
-(Document(page_content="foo", metadata={"page": "0"}), 0.7071067811865475),
+(Document(page_content="foo", metadata={"page": "0"}), 1.0),
```


Checked test, it must be 1 indeed => we have same value -> vectors coincide -> return 1 which is max similarity.
docsearch.similarity_search_with_relevance_scores("foo", k=3) and page_content="foo" are same

amotl replied:
> looks like we had exactly one instance of "1 will become 0.714"

Exactly, your admonition was right on the spot. Thanks again!

amotl committed:
CrateDB: Vector Store -- make _euclidean_relevance_score_fn identity f.

We don't need anything on top of it, ie we don't need this function and
instead should use value from CrateDB as is.

Similarity is already in the (0,1] interval and dividing by math.sqrt(2)
won't normalize it but return wrong result, for example 1 will become
0.714.

BaurzhanSakhariev reviewed (approved):
Thanks!

amotl commented:
Hi @eyurtsev. We think our patches are ready to be merged. This one, and GH-27711 as well as GH-27712. May we humbly ask you to have a look?

kneth commented on libs/community/langchain_community/vectorstores/cratedb/base.py:
```diff
@@ -0,0 +1,474 @@
+        # (0 is most similar, sqrt(2) most dissimilar)
+        # to a similarity function (0 to 1)
+
+        # Original:
```
Should we remove this and next line?

amotl replied:
Thanks. 4b9310e517 improves that section, removing the lines you flagged, and replacing them with a better description about CrateDB's implementation.

kneth reviewed (approved):
LGTM

amotl committed:
CrateDB: Vector Store -- update docs in _euclidean_relevance_score_fn

amotl commented:
Dear @eyurtsev,

may I humbly ask you if you could afford a few cycles to review our patches? Thanks in advance!

With kind regards,
Andreas.

efriis commented:
Hey! This adds a net-new community integration or feature, which has been replaced by dedicated integration packages. I'll close this PR if it's ok with you, and would recommend reopening with just docs updates, as well as registering your package in `libs/packages.yml`! We'll be able to review simple PRs that only modify these two things much faster :)

Here's the guide, and if you have questions, feel free to leave them in the comments on those pages so others can see them! https://python.langchain.com/docs/contributing/how_to/integrations/

This will pair very nicely with the variety of integrations you're working on at the moment! 

Will leave this PR open to discuss and link back here when closing the other ones (to make sure we're discussing in one place)

amotl commented:
Hi Erick,

thanks for your reply. So, we will conceive and publish a dedicated Python package `langchain-cratedb` then, including the relevant code? Sure, we can do.

> Will leave this PR open to discuss and link back here when closing the other ones (to make sure we're discussing in one place)

I think it would be coherent to also close this PR, and then discuss on behalf of a separate dedicated **issue** to accompany the genesis of `langchain-cratedb`, where I am very much looking forward to, receiving your assistance on any question we may have along the way. 🍀

I will open the other issue when it is time to start the discussion, i.e. when we have something to show that starts working. Do you agree with this approach?

With kind regards,
Andreas.

efriis commented:
sounds like a great approach!
